src/main/scala/peregin/gpv/model/TrackPoint.scala

```diff
@@ -43,14 +44,17 @@ case class TrackPoint(position: GeoPosition,
   def analyze(next: TrackPoint, prevs: Seq[TrackPoint]): Unit = {
     segment = distanceTo(next)
     next.distance = distance + segment
-    val dt = (next.time.getMillis - time.getMillis).toDouble / TrackPoint.millisToHours
-    if (dt != 0d) speed = segment / dt
-    // smoother grade calculation, consider the previous point as well, otherwise some outliers are being introduced
-    val (gradeSegment, firstElevation) = {
-      if (prevs.isEmpty) (segment, elevation)
-      else (prevs.map(_.segment).sum + segment, prevs.head.elevation)
-    }
-    if (gradeSegment > 0) grade = (next.elevation - firstElevation) / (gradeSegment * 10)
+
+    // smoother speed calculation, consider up to 3 seconds back and 1 point forward:
+    val firstSpeed: TrackPoint = SeqUtil.findReverseTill(prevs, (p: TrackPoint) => this.time.getMillis - p.time.getMillis <= 3000).getOrElse(this)
+    val dt = (next.time.getMillis - firstSpeed.time.getMillis).toDouble / TrackPoint.millisToHours
+    if (dt != 0d) speed = (next.distance - firstSpeed.distance) / dt
+
+    // smoother grade calculation, consider up to 9 seconds back and 1 point forward:
+    val firstElevation: TrackPoint = SeqUtil.findReverseTill(prevs, (p: TrackPoint) => this.time.getMillis - p.time.getMillis <= 9000).getOrElse(this)
+    val distanceElevation = next.distance - firstElevation.distance
+    if (distanceElevation > 0) grade = (next.elevation - firstElevation.elevation) / distanceElevation * (100 / 1000.0)
```

Comment:
last part of the expression could be simplified/replaced `(100 / 1000.0)` with `0.1`

Reply:
I know, @peregin but want to make it more visible for humans (compiler optimizes it anyway).  I was confused about the previous constant, until it showed huge numbers and I realized the distance was on kilometers :-D